synapseclient/core/upload/multipart_upload.py

```diff
@@ -3,385 +3,526 @@
 Synapse Multipart Upload
 ************************
 
-Implements the client side of `Synapse multipart upload`_, which provides a robust means of uploading large files (into
-the 10s of GB). End users should not need to call any of these functions directly.
+Implements the client side of `Synapse multipart upload`_, which provides a
+robust means of uploading large files (into the 10s of GB). End users should
+not need to call any of these functions directly.
 
 .. _Synapse multipart upload:
  http://docs.synapse.org/rest/index.html#org.sagebionetworks.file.controller.UploadController
 
 """
 
+import concurrent.futures
+from collections import OrderedDict
 import hashlib
 import json
 import math
 import mimetypes
+import multiprocessing
 import os
 import requests
+import threading
 import time
-import warnings
-import ctypes
+from typing import List
 
-from synapseclient.core import pool_provider, exceptions
+from synapseclient.core import exceptions, pool_provider
 from synapseclient.core.utils import printTransferProgress, md5_for_file, MB
-from synapseclient.core.models.dict_object import DictObject
-from synapseclient.core.exceptions import SynapseError
-from synapseclient.core.exceptions import SynapseHTTPError
-from synapseclient.core.utils import threadsafe_generator
 
+# AWS limits
 MAX_NUMBER_OF_PARTS = 10000
-MIN_PART_SIZE = 8*MB
-MAX_RETRIES = 7
-
-
-def find_parts_to_upload(part_status):
-    """
-    Given a string of the form "1001110", where 1 and 0 indicate a status of completed or not, return the part numbers
-    that aren't completed.
-    """
-    return [i+1 for i, c in enumerate(part_status) if c == '0']
-
+MIN_PART_SIZE = 5 * MB
 
-def count_completed_parts(part_status):
-    """
-    Given a string of the form "1001110", where 1 and 0 indicate a status of completed or not, return the count of
-    parts already completed.
-    """
-    return len([c for c in part_status if c == '1'])
+# default number of threads to use if none specified.
+# default may be adjusted down depending on the specifics of the upload.
+DEFAULT_MAX_WORKERS = 16
 
+# ancient tribal knowledge
+DEFAULT_PART_SIZE = 8 * MB
+MAX_RETRIES = 7
 
-def calculate_part_size(fileSize, partSize=None, min_part_size=MIN_PART_SIZE, max_parts=MAX_NUMBER_OF_PARTS):
-    """
-    Parts for multipart upload must be at least 5 MB and there must be at most 10,000 parts
-    """
-    if partSize is None:
-        partSize = max(min_part_size, int(math.ceil(fileSize/float(max_parts))))
-    if partSize < min_part_size:
-        raise ValueError('Minimum part size is %d MB.' % (min_part_size/MB))
-    if int(math.ceil(float(fileSize) / partSize)) > max_parts:
-        raise ValueError('A part size of %0.1f MB results in too many parts (%d).'
-                         % (float(partSize)/MB, int(math.ceil(fileSize / partSize))))
-    return partSize
 
+thread_local = threading.local()
+
+
+class UploadAbortedException(Exception):
```

Comment:
Should these be in synapseclient.core.exceptions?

Reply:
Good point.